_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,894 @@
+
+Here's the full code
+
+```scala
+package com.rockthejvm
+
+import scala.util.*
+import java.net.URLEncoder
+import com.google.zxing.common.BitMatrix
+import com.google.zxing.MultiFormatWriter
+import com.google.zxing.BarcodeFormat
+import java.io.FileOutputStream
+import com.google.zxing.client.j2se.MatrixToImageWriter
+
+trait BarCodeService:
+    def getGoogleAuthenticatorBarCode(secretKey: String, account: String, issuer: String, generator: String): Try[String] =
+        generator match
+            case "Hotp" =>
+                Try{
+                s"otpauth://${generator.toLowerCase()}/" +
+                URLEncoder.encode(issuer + ":" + account, "UTF-8").replace("+","%20") +
+                "?secret=" + URLEncoder.encode(secretKey, "UTF-8").replace("+","%20") +
+                "&issuer=" + URLEncoder.encode(issuer, "UTF-8").replace("+","%20") +
+                "&conter=0"
+                }
+            case "Totp" =>
+                Try{
+                    s"otpauth://${generator.toLowerCase()}/" +
+                    URLEncoder.encode(issuer + ":" + account, "UTF-8").replace("+","%20") +
+                    "?secret=" + URLEncoder.encode(secretKey, "UTF-8").replace("+","%20") +
+                    "&issuer=" + URLEncoder.encode(issuer, "UTF-8").replace("+","%20")
+                }
+
+    def createQRCode(barCodeData: String, filePath: String, height: Int, width: Int) =
+        val matrix: BitMatrix = new MultiFormatWriter().encode(barCodeData, BarcodeFormat.QR_CODE, width, height)
+        val outVal: FileOutputStream = new FileOutputStream(filePath)
+        MatrixToImageWriter.writeToStream(matrix,"png",outVal)
+end BarCodeService
+```
+
+### 3.1.2 Creating the OtpService
+
+Let's create a scala file and save it as `OtpService.scala`. This service will contain the implementation for `Hotp` and `Totp` that we will use for this application.
+
+```scala
+package com.rockthejvm
+
+import com.bastiaanjansen.otp.*
+
+class OtpService(generator: String, id: Int) extends MyDatabase with BarCodeService:
+    private val secret: Array[Byte] = SecretGenerator.generate()
+end OtpService
+```
+
+We create it as a class in order to pass some arguments to the constructor during instantiation. As mentioned before, `generator` is a string of either `Hotp` or `Totp` while `id` points to a user in the database. This class extends `MyDatabase` and `BarCodeService` traits since we will make use of some of their functions.
+
+We first create our `secret` as a private val using `otp-java`, since we will only be using it within this class
+
+```scala
+import java.time.Duration
+
+class OtpService(generator: String, id: Int) extends MyDatabase with BarCodeService:
+    ...
+    enum Generator:
+        case Hotp, Totp
+
+        def getHotp: HOTPGenerator = new HOTPGenerator.Builder(secret)
```

Comment:
This seems a bit off - you're returning a new generator every time the method is called, and a `Hotp` can call `getTotp` and viceversa.

You might do better by creating a sealed trait Generator and have Hotp and Totp as objects that extend that, so that everything is clear and separated, and have the Hotp and Totp as singletons so that we don't proliferate all these instances.

Reply:
fixed this